### ISSUE TITLE
Crash with auth_user when connecting to the admin database

### ISSUE THREAD

**Issue by PierreDucroquet** [pgbouncer/pgbouncer#568]
Hello

It seems 383af567605fadceb351833ac346313500f13687 introduced a severe regression. Making auth_user global that way made it possible to crash pgbouncer by simply doing `psql -p 6432 -U does_not_exists pgbouncer`
When this is done, pgbouncer writes in its log the following line and exits:
```
2021-02-11 11:08:25.338 UTC [22164] FATAL @src/objects.c:312 in function put_in_order(): put_in_order: found existing elem
```

**Comment by petere** [comment]
I couldn't reproduce this. Presumably, it depends on some details in the configuration. Can you share your full configuration files?

**Comment by pavel-tarasov** [comment]
Same problem, maybe my setting will help you to represent this bug:
PgBouncer version 1.15.0
**pgbouncer.ini:**
```
[databases]
postgres = host=localhost port=5432 dbname=postgres
daily_candles = host=localhost port=5432 dbname=daily_candles
history = host=localhost port=5432 dbname=history
history_ts = host=localhost port=5432 dbname=history_ts
ldb_data = host=localhost port=5432 dbname=ldb_data
metadata = host=localhost port=5432 dbname=metadata
research = host=localhost port=5432 dbname=research
trades = host=localhost port=5432 dbname=trades
[users]
[pgbouncer]
logfile = /var/log/postgresql/pgbouncer.log
pidfile = /var/run/postgresql/pgbouncer.pid
listen_addr = *
listen_port = 6432
unix_socket_dir = /var/run/postgresql
auth_type = md5
auth_file = /etc/pgbouncer/userlist.txt
auth_query = SELECT * FROM public.user_lookup($1);
auth_user = pgbouncer
admin_users = pgbouncer
pool_mode = transaction
ignore_startup_parameters = extra_float_digits
application_name_add_host = 1
max_client_conn = 1000
default_pool_size = 20
reserve_pool_size = 10
max_db_connections = 40
log_connections = 1
log_disconnections = 1
```
I run:
`psql -h localhost -p 6432 -d pgbouncer -U nonexistinguser`
And get in log:
`FATAL @src/objects.c:312 in function put_in_order(): put_in_order: found existing elem`

This problem can be solved by setting auth_user for each DB in [databases] section, but even in this case attempts to connect to pgbouncer DB fail with `psql: error: could not connect to server: ERROR:  password authentication failed` for any user that doesn't exist in userlist.txt file. If user exists in the userlist.txt file attempt to connect will fail with the correct understandable error "not allowed". Probably this is just improvement, not a bug, but maybe it will be more correct to return "not allowed" if username not in admin or stat users, no matter where we got username, from userlist.txt or from auth_query?

**Comment by petere** [comment]
Thanks, I was able to reproduce it. You just need

    auth_user = pgbouncer

I think this is essentially the same bug as #90 and #314. The reason this wasn't seen before is that in previous versions, the global `auth_user` setting was essentially ignored, but now it's not. Because of the way the admin database is set up, you get one global user named "pgbouncer" and one forced user "pgbouncer", and they conflict with each other, leading to this error message. If you squint, this is exactly the same problem as described in #90, except that pgbouncer triggers it internally.

The workaround is to not use "pgbouncer" as the name for the global `auth_user`.

**Comment by scorpevans** [comment]
I'm currently using `auth_user=pgbouncer` without any issues.

Could the issue be with the `auth_query` returning empty rows?
The pgbouncer doc gives an auth_query which `RETURNS record` which crashes pgbouncer; using something to the effect of `RETURNS SETOF record` seems to solve this:

```
-- when the user is not in auth_file, it is looked up in auth_query
postgres=# select * from user_lookup('non-existing-user');
 uname | phash 
-------+-------
       | 
(1 row)

-- the lookup must not return an empty row
postgres=# select * from user_lookup('non-existing-user');
 uname | phash 
-------+-------
(0 rows)
```

**Comment by JelteF** [comment]
This should be fixed by #1052
